cpp/ql/lib/semmle/code/cpp/ir/implementation/aliased_ssa/gvn/internal/ValueNumberingInternal.qll

```diff
@@ -92,6 +92,20 @@ predicate callArgValueNumber(CallInstruction call, int index, TValueNumber arg)
 predicate callArgRank(CallInstruction call, int index, Instruction arg) {
   arg =
     rank[index](int argIndex, boolean isEffect, Instruction instr |
+      exists(CallSideEffectInstruction cse |
+        cse.getPrimaryInstruction() = call and
+        cse.getSideEffectOperand().getAnyDef() = instr and
+        argIndex = -2 and
+        isEffect = false
+      )
+      or
+      exists(CallReadSideEffectInstruction cse |
+        cse.getPrimaryInstruction() = call and
+        cse.getSideEffectOperand().getAnyDef() = instr and
+        argIndex = -2 and
+        isEffect = false
+      )
+      or
```

Comment:
It probably doesn't matter a lot, but shouldn't both of these be `isEffect = true`?

Reply:
Alternatively, we could choose to only GVN functions that are side-effect free (which I think would cover `std::vector`'s `size` function, right?)

Reply:
I'd prefer to GVN all functions if we can, and eventually GVN the memory side effects as well (but that's work for another PR)